test/unit/diff.hpp

```diff
@@ -0,0 +1,201 @@
+                // This step ensures that the table holds the length of the longest common
+                // subsequence found so far.
+                lcsTable[i + 1][j + 1] = std::max(lcsTable[i + 1][j], lcsTable[i][j + 1]);
+            }
+        }
+    }
+
+    // Traceback to find the differences
+    DiffStringsResult result;
+    struct DiffLineResult {
+        std::string line;
+        bool added{false};
+        bool removed{false};
+    };
+    std::vector<DiffLineResult> diffLines;
+    size_t i = lines1.size();
+    size_t j = lines2.size();
+
+    // Traverse the LCS table
+    // The algorithm starts in the bottom right corner of the table
+    // Starting from the bottom-right cell of the LCS table, it examines
+    // the adjacent cells to determine the direction of the LCS
+    while (i > 0 && j > 0) {
+        if (lines1[i - 1] == lines2[j - 1]) {
+            // If the current lines lines1[i-1] and lines2[j-1] are equal,
+            // it means the line is common to both multiline strings. It
+            // is added to diffLines with a space prefix, indicating no change.
+            diffLines.push_back({std::string(lines1[i - 1]), false, false});
+            --i;
+            --j;
+            result.unchanged++;
+        } else if (lcsTable[i][j - 1] >= lcsTable[i - 1][j]) {
+            // If the value in the cell on the left, lcsTable[i][j-1], is
+            // greater than or equal to the value in the cell above,
+            // lcsTable[i-1][j], it means the line in lines2[j-1] is
+            // part of the LCS. Thus, it is added to diffLines with
+            // a "+" prefix to indicate an addition.
+            diffLines.push_back({std::string(lines2[j - 1]), true, false});
+            --j;
+            result.added++;
+        } else {
+            // Otherwise, the line in lines1[i-1] is part of the LCS, and it
+            // is added to diffLines with a "-" prefix to indicate a deletion.
+            diffLines.push_back({std::string(lines1[i - 1]), false, true});
+            --i;
+            result.removed++;
+        }
+    }
+
+    while (i > 0) {
+        diffLines.push_back({std::string(lines1[i - 1]), false, true});
+        --i;
+        result.removed++;
+    }
+
+    while (j > 0) {
+        diffLines.push_back({std::string(lines2[j - 1]), true, false});
+        --j;
+        result.added++;
+    }
+
+    // Reverse the diff lines to match the original order
+    std::reverse(diffLines.begin(), diffLines.end());
+
+    // Concatenate diff lines into a single string considering number of unchanged lines in the context
+    std::size_t unchanged = 0;
+    auto last_rendered = std::size_t(-1);
+    for (i = 0; i < diffLines.size(); ++i) {
+        auto& diffLine = diffLines[i];
+        if (diffLine.added || diffLine.removed) {
+            std::size_t context_begin = std::max(i - context_size, (last_rendered == std::size_t(-1) ? 0 : last_rendered + 1));
```

Comment:
You aren't protecting `max` from Windows macro shenanigans

Reply:
Oh... I thought that wasn't a thing anymore.